File: templates/ruflet_flutter_template/lib/ruflet_spinkit.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';

class RufletSpinKitExtension extends FletExtension {
  @override
  Widget? createWidget(Key? key, Control control) {
    if (control.type != 'RufletSpinKit') return null;
    return RufletSpinKitControl(key: key, control: control);
  }
}

class RufletSpinKitControl extends StatelessWidget {
  const RufletSpinKitControl({super.key, required this.control});

  final Control control;

  @override
  Widget build(BuildContext context) {
    final color =
        control.getColor('color', context) ??
        Theme.of(context).colorScheme.primary;
    final size = control.getDouble('size', 50)!;
    final duration = Duration(milliseconds: control.getInt('duration', 1200)!);
    final variant = control.getString('variant', 'rotating_circle')!;

    return LayoutControl(
      control: control,
      child: _spinner(variant, color, size, duration),
    );
  }

  Widget _spinner(String variant, Color color, double size, Duration duration) {
    switch (variant) {
      case 'rotating_plain':
        return SpinKitRotatingPlain(
          color: color,
          size: size,
          duration: duration,
        );
      case 'double_bounce':
        return SpinKitDoubleBounce(
          color: color,
          size: size,
          duration: duration,
        );
      case 'wave':
        return SpinKitWave(color: color, size: size, duration: duration);
      case 'wandering_cubes':
        return SpinKitWanderingCubes(
          color: color,
          size: size,
          duration: duration,
        );
      case 'fading_four':
        return SpinKitFadingFour(color: color, size: size, duration: duration);
      case 'fading_cube':
        return SpinKitFadingCube(color: color, size: size, duration: duration);
      case 'pulse':
        return SpinKitPulse(color: color, size: size, duration: duration);
      case 'chasing_dots':
        return SpinKitChasingDots(color: color, size: size, duration: duration);
      case 'three_bounce':
        return SpinKitThreeBounce(color: color, size: size, duration: duration);
      case 'circle':
        return SpinKitCircle(color: color, size: size, duration: duration);
      case 'cube_grid':
        return SpinKitCubeGrid(color: color, size: size, duration: duration);
      case 'fading_circle':
        return SpinKitFadingCircle(
          color: color,
          size: size,
          duration: duration,
        );
      case 'folding_cube':
        return SpinKitFoldingCube(color: color, size: size, duration: duration);
      case 'pumping_heart':
        return SpinKitPumpingHeart(
          color: color,
          size: size,
          duration: duration,
        );
      case 'hour_glass':
        return SpinKitHourGlass(color: color, size: size, duration: duration);
      case 'pouring_hour_glass':
        return SpinKitPouringHourGlass(
          color: color,
          size: size,
          duration: duration,
        );
      case 'pouring_hour_glass_refined':
        return SpinKitPouringHourGlassRefined(
          color: color,
          size: size,
          duration: duration,
        );
      case 'fading_grid':
        return SpinKitFadingGrid(color: color, size: size, duration: duration);
      case 'ring':
        return SpinKitRing(color: color, size: size, duration: duration);
      case 'ripple':
        return SpinKitRipple(color: color, size: size, duration: duration);
      case 'dual_ring':
        return SpinKitDualRing(color: color, size: size, duration: duration);
      case 'spinning_circle':
        return SpinKitSpinningCircle(
          color: color,
          size: size,
          duration: duration,
        );
      case 'spinning_lines':
        return SpinKitSpinningLines(
          color: color,
          size: size,
          duration: duration,
        );
      case 'square_circle':
        return SpinKitSquareCircle(
          color: color,
          size: size,
          duration: duration,
        );
      case 'three_in_out':
        return SpinKitThreeInOut(color: color, size: size, duration: duration);
      case 'dancing_square':
        return SpinKitDancingSquare(
          color: color,
          size: size,
          duration: duration,
        );
      case 'piano_wave':
        return SpinKitPianoWave(color: color, size: size, duration: duration);
      case 'pulsing_grid':
        return SpinKitPulsingGrid(color: color, size: size, duration: duration);
      case 'wave_spinner':
        return SpinKitWaveSpinner(color: color, size: size, duration: duration);
      case 'rotating_circle':
      default:
        return SpinKitRotatingCircle(
          color: color,
          size: size,
          duration: duration,
        );
    }
  }
}

File: templates/ruflet_flutter_template/apple_packages/ruflet_apple/Sources/RufletExtensions/RufletSpinKit/RufletSpinKit.swift
public enum RufletSpinKit {
  public static let packageName = "ruflet_spinkit"
  public static let canonicalControlType = "RufletSpinKit"

  /// Canonical Ruflet variants mapped to the exact pinned Flet control type.
  public static let variantControlTypes: [String: String] = [
    "rotating_circle": "SpinKitRotatingCircle",
    "rotating_plain": "SpinKitRotatingPlain",
    "double_bounce": "SpinKitDoubleBounce",
    "wave": "SpinKitWave",
    "wandering_cubes": "SpinKitWanderingCubes",
    "fading_four": "SpinKitFadingFour",
    "fading_cube": "SpinKitFadingCube",
    "pulse": "SpinKitPulse",
    "chasing_dots": "SpinKitChasingDots",
    "three_bounce": "SpinKitThreeBounce",
    "circle": "SpinKitCircle",
    "cube_grid": "SpinKitCubeGrid",
    "fading_circle": "SpinKitFadingCircle",
    "folding_cube": "SpinKitFoldingCube",
    "pumping_heart": "SpinKitPumpingHeart",
    "hour_glass": "SpinKitHourGlass",
    "pouring_hour_glass": "SpinKitPouringHourGlass",
    "pouring_hour_glass_refined": "SpinKitPouringHourGlassRefined",
    "fading_grid": "SpinKitFadingGrid",
    "ring": "SpinKitRing",
    "ripple": "SpinKitRipple",
    "dual_ring": "SpinKitDualRing",
    "spinning_circle": "SpinKitSpinningCircle",
    "spinning_lines": "SpinKitSpinningLines",
    "square_circle": "SpinKitSquareCircle",
    "three_in_out": "SpinKitThreeInOut",
    "dancing_square": "SpinKitDancingSquare",
    "piano_wave": "SpinKitPianoWave",
    "pulsing_grid": "SpinKitPulsingGrid",
    "wave_spinner": "SpinKitWaveSpinner",
  ]

  public static let pinnedControlTypes: Set<String> = [
    "SpinKitRotatingPlain", "SpinKitDoubleBounce", "SpinKitWave", "SpinKitWanderingCubes",
    "SpinKitFadingFour", "SpinKitFadingCube", "SpinKitPulse", "SpinKitChasingDots",
    "SpinKitThreeBounce", "SpinKitCircle", "SpinKitCubeGrid", "SpinKitFadingCircle",
    "SpinKitRotatingCircle", "SpinKitFoldingCube", "SpinKitPumpingHeart", "SpinKitHourGlass",
    "SpinKitPouringHourGlass", "SpinKitPouringHourGlassRefined", "SpinKitFadingGrid",
    "SpinKitRing", "SpinKitRipple", "SpinKitDualRing", "SpinKitSpinningCircle",
    "SpinKitSpinningLines", "SpinKitSquareCircle", "SpinKitThreeInOut",
    "SpinKitDancingSquare", "SpinKitPianoWave", "SpinKitPulsingGrid", "SpinKitWaveSpinner",
  ]

  /// Exact types claimed by the extension registry. Keep this a literal set so
  /// the source auditor can prove each dispatch edge without executing Swift.
  public static let controlTypes: Set<String> = [
    "RufletSpinKit", "SpinKitRotatingPlain", "SpinKitDoubleBounce", "SpinKitWave",
    "SpinKitWanderingCubes", "SpinKitFadingFour", "SpinKitFadingCube", "SpinKitPulse",
    "SpinKitChasingDots", "SpinKitThreeBounce", "SpinKitCircle", "SpinKitCubeGrid",
    "SpinKitFadingCircle", "SpinKitRotatingCircle", "SpinKitFoldingCube",
    "SpinKitPumpingHeart", "SpinKitHourGlass", "SpinKitPouringHourGlass",
    "SpinKitPouringHourGlassRefined", "SpinKitFadingGrid", "SpinKitRing", "SpinKitRipple",
    "SpinKitDualRing", "SpinKitSpinningCircle", "SpinKitSpinningLines", "SpinKitSquareCircle",
    "SpinKitThreeInOut", "SpinKitDancingSquare", "SpinKitPianoWave", "SpinKitPulsingGrid",
    "SpinKitWaveSpinner",
  ]

  public static func resolvedControlType(controlType: String, variant: String?) -> String? {
    if controlType == canonicalControlType {
      guard let variant else { return nil }
      return variantControlTypes[variant]
    }
    return pinnedControlTypes.contains(controlType) ? controlType : nil
  }
}
